src/module_wrap.cc

```diff
@@ -761,7 +761,11 @@ Maybe<URL> FinalizeResolution(Environment* env,
     return Nothing<URL>();
   }
 
+#ifdef _WIN32
+  const std::string& path = "\\\\.\\" + resolved.ToFilePath();
```

Comment:
There are quite a few other places where we use `resolved.ToFilePath()` in the resolver and expect to be able to run descriptor checks against that.

I'm wondering if it might make more sense to implement a `ToSafeFilePath()` here on the URL side, then update these calls to that? That way we get this feature for our package boundary checks as well.

Eg I think that `C:\path\to\con\package.json` would likely still throw if I'm understanding the issue correctly, even with this PR merged.

Reply:
Are there downsides to always use this prefix in `ToFilePath()`? Readability concerns?

Reply:
The concern would be when the output is directly passed to JS, since the expectation there is that Windows paths don't start with `\\`. But perhaps there are no places where this API is used that way...

Reply:
I updated `OpenDescriptor` to perform the check and added a test to ensure that `package.json` works as well. Now, If the resolution fails, the error that is reported to JS does not include prefix in the path. Is there still a use-case where this does not work?

Reply:
Thanks for checking this - it seems to only apply to these file checks then.

We do have file checks in another place and that is for extension searching on CommonJS packages. So a CommonJS package containing:

```json
{
  "main": "con"
}
```

will do extension checking checking first for `con` which would throw. So to be complete we should add all these cases too as in the `FileExists` function at https://github.com/nodejs/node/blob/master/src/module_wrap.cc#L669.

Reply:
I'm not sure what exactly you mean... Can you be more precise, or create a gist with the changes you propose?

> So a CommonJS package containing `{"main":"con"}` will do extension checking checking first for con which would throw.

I'm unable to reproduce that.

Reply:
I just tested this out with - 

```
node_modules/pkg/package.json: { "main": "con" }
node-modules/pkg/con.js: module.exports = 'blah'
x.mjs: import pkg from 'pkg';
```

And I can verify it's failing the main lookup under `--experimental-modules` when any other filename will succeed.

As mentioned just the `FileExists` modifications as above should fix this.

Reply:
So, this does not work for you?

```console
D:\Git\node\node> git log --format="%H" -n 1
7ee23336b3020b513b247dcc678c3aa960748b79

D:\Git\node\node> cd ..

D:\Git\node> type "\\.\D:\Git\node\x.mjs"
import pkg from 'pkg';
console.log(pkg);

D:\Git\node> type "\\.\D:\Git\node\node_modules\pkg\package.json"
{ "main": "con" }

D:\Git\node> type "\\.\D:\Git\node\node_modules\pkg\con.js"
module.exports = 'blah';

D:\Git\node> ".\node\Release\node.exe" --experimental-modules x.mjs
(node:2856) ExperimentalWarning: The ESM module loader is experimental.
blah
```

Reply:
Yes, although I tested against the current branch not your PR. If this is working on the current branch then that is all the cases sorted - sorry for confusion!